libpod/pod.go

```diff
@@ -111,77 +103,85 @@ func (p *Pod) Start() (map[string]error, error) {
-				if depCtr.state.State != ContainerStateRunning &&
-					depCtr.state.State != ContainerStatePaused {
-					// We are definitely not OK to init, a dependency is not up
-					depsOK = false
-					// Check to see if the dependency errored
-					// If it did, error here too
-					if _, ok := ctrErrors[depID]; ok {
-						depErr = errors.Wrapf(ErrCtrStateInvalid, "dependency %s of container %s failed to start", depID, id)
-					}
-
-					break
-				}
-			}
-			if !depsOK {
-				// Only if one of the containers dependencies failed should we stop trying
-				// Otherwise, assume it's just yet to start, retry starting this container later
-				if depErr != nil {
-					ctrErrors[id] = depErr
-					delete(ctrsToStart, id)
-				}
-				continue
-			}
+	// If there are no containers without dependencies, we can't start
+	// Error out
+	if len(graph.noDepNodes) == 0 {
+		return nil, errors.Wrapf(ErrNoSuchCtr, "no containers in pod %s have no dependencies, cannot start pod", p.ID())
+	}
 
-			// Initialize and start the container
-			if err := ctr.initAndStart(); err != nil {
-				ctrErrors[id] = err
-			}
-			delete(ctrsToStart, id)
+	// Traverse the graph beginning at nodes with no dependencies
+	for _, node := range graph.noDepNodes {
+		startNode(node, false, ctrErrors, ctrsVisited)
+	}
+
+	return ctrErrors, nil
+}
+
+// Visit a node on a container graph and start the container, or set an error if
+// a dependency failed to start
+func startNode(node *containerNode, setError bool, ctrErrors map[string]error, ctrsVisited map[string]bool) {
+	// First, check if we have already visited the node
+	if ctrsVisited[node.id] {
+		return
+	}
+
+	// If setError is true, a dependency of us failed
+	// Mark us as failed and recurse
+	if setError {
+		// Mark us as visited, and set an error
+		ctrsVisited[node.id] = true
+		ctrErrors[node.id] = errors.Wrapf(ErrCtrStateInvalid, "a dependency of container %s failed to start", node.id)
+
+		// Hit anyone who depends on us, and set errors on them too
+		for _, successor := range node.dependedOn {
+			startNode(successor, true, ctrErrors, ctrsVisited)
 		}
 
-		loopCounter = loopCounter - 1
-		if loopCounter == 0 {
-			// Loop through all remaining containers and add an error
-			for id := range ctrsToStart {
-				ctrErrors[id] = errors.Wrapf(ErrInternal, "exceeded maximum attempts trying to start container %s", id)
-			}
+		return
+	}
 
-			break
+	// Have all our dependencies started?
+	// If not, don't visit the node yet
+	depsVisited := true
+	for _, dep := range node.dependsOn {
+		depsVisited = depsVisited && ctrsVisited[dep.id]
+	}
+	if !depsVisited {
+		// Don't visit us yet, all dependencies are not up
+		// We'll hit the dependencies eventually, and when we do it will
+		// recurse here
+		return
+	}
+
+	// Going to start the container, mark us as visited
+	ctrsVisited[node.id] = true
+
+	// Start the container (only if it is not running)
+	ctrErrored := false
+	if ctr.state.State != ContainerStateRunning {
```

Comment:
I don't think ctr is defined anymore.  

# github.com/projectatomic/libpod/libpod
libpod/pod.go:172: undefined: ctr in ctr.state

Reply:
Oops. Fixed.